### PR TITLE
Ensure ListResources is accurate with denied resources

### PR DESCRIPTION
**Pull request by rosstimothy**
Prior to this tsh ls and the web UI were reporting different number
of resources when a user had role/roles that restricted access to
certain resources. This was caused by returning an incorrect
`NextKey` in the `ListResourcesResponse`.

In the event that `IterateResourcePages` was returning early due to
the callback requesting to stop iteration, the `NextKey` being returned
was the last key for the **current page** instead of the key for
the **next resource** in the page. This would cause N resources between
the last resource iterated on and the end of the page to be skipped.

To remedy this `IterateResourcePages` has been refactored to
`IterateResources`. This shifts the responsiblity of building the
`ListResourcesResponse` onto the caller and not the iterator. There
are no more pages to worry about, so the logic about what the next
key should be becomes simpler.

`TestListResources_WithRoles` was added to ensure that a user with
a set of roles that denies access is always returned the correct
resources from `ListResources` from both tsh and the web ui.

### CODE REVIEW

File: lib/auth/auth_with_roles.go [hunk 1]
```diff
@@ -1135,45 +1136,42 @@ func (a *ServerWithRoles) ListResources(ctx context.Context, req proto.ListResou
...
-				break
-			}
-
-			if err := resourceChecker.CanAccess(resource); err != nil {
-				if trace.IsAccessDenied(err) {
-					continue
-				}
+	resp := &types.ListResourcesResponse{
+		Resources: make([]types.ResourceWithLabels, 0, limit),
+	}
```
Comment (fspmarshall) [inline lib/auth/auth_with_roles.go:1146]: I think we should cap the amount of preallocation we're willing to do.  Wouldn't want someone to be able to deliberately OOM the auth server by sending requests with giant limit values.

File: lib/auth/auth.go [hunk 2]
```diff
@@ -2852,11 +2852,13 @@ func (a *Server) GetNodes(ctx context.Context, namespace string) ([]types.Server
...
 
-// ResourcePageFunc is a function to run on each page iterated over.
-type ResourcePageFunc func(next []types.ResourceWithLabels) (stop bool, err error)
+// ErrDone indicates that resource iteration is complete
+var ErrDone = errors.New("done iterating")
 
-// IterateResourcePages can be used to iterate over pages of resources.
-func (a *Server) IterateResourcePages(ctx context.Context, req proto.ListResourcesRequest, f ResourcePageFunc) (*types.ListResourcesResponse, error) {
+// IterateResources will pass all resources matching the request to the provided function f until it has either
+// exhausted all matching resources, or f returns ErrDone. When ErrDone is provided nil will be returned, otherwise
```
Comment (zmb3) [inline lib/auth/auth.go:2859]: What does `when ErrDone is provided` mean?

**Review by zmb3 — APPROVED** [review]

File: lib/auth/auth_with_roles.go (hunk 1 above)
Comment (espadolini) [reply lib/auth/auth_with_roles.go:1146]: I was going to say something about it only messing up some GC stats, but at least on macOS a `make([]byte, 0, 1<<40)` does seem to actually try to zero out a TiB of ram 😔

Should we just not preallocate? The usual reallocation strategy should get to an appropriate size pretty quickly even starting from 0.

**Review by espadolini — APPROVED** [review]
LGTM but the fact that the limit asked for by the client ends up being the limit used for the internal pagination seems a bit fishy

**Pull request merged**